### PR TITLE
core(preload): only allow same origin (domain + subdomains)

### PR DESCRIPTION
**Pull request by wardpeet**
Fixes nr 1 of #4425 
Restrict results to same-origin (as a rough filter for predictable and unchanging URLs). Same origin as in subdomains and rootdomains.

### CODE REVIEW

**Comment by patrickhulce** [comment]
@wardpeet FYI I had to rework considerable portions of this for lantern (#5062), wdyt about holdin' tight for a cycle to get that in first? :)

**Comment by wardpeet** [comment]
sure thing no worries

File: lighthouse-core/lib/url-shim.js [hunk 1]
```diff
@@ -87,6 +87,26 @@ class URLShim extends URL {
     }
   }
 
+  /**
+   * Check if rootDomains matches
+   *
+   * @param {string} urlA
+   * @param {string} urlB
+   */
+  static rootDomainsMatch(urlA, urlB) {
```
Comment (paulirish) [inline lighthouse-core/lib/url-shim.js]: can you add some tests in url-shim-test?

File: lighthouse-core/lib/url-shim.js [hunk 2]
```diff
@@ -87,6 +87,26 @@ class URLShim extends URL {
...
+    const urlAInfo = new URL(urlA);
+    const urlBInfo = new URL(urlB);
+
+    if (!urlAInfo.hostname || !urlBInfo.hostname) {
+      return false;
+    }
+
+    const urlARootDomain = urlAInfo.hostname.split('.').slice(-2).join('.');
```
Comment (paulirish) [inline lighthouse-core/lib/url-shim.js]: this is called "tld plus one". let's use that term for the method name and here.

let's add a few whitelisted exceptions like '.co.uk'. (country-code second level domains).

File: lighthouse-core/lib/url-shim.js [hunk 3]
```diff
@@ -87,6 +87,26 @@ class URLShim extends URL {
...
+    const urlBRootDomain = urlBInfo.hostname.split('.').slice(-2).join('.');
```
Comment (patrickhulce) [inline lighthouse-core/lib/url-shim.js]: https://en.wikipedia.org/wiki/Second-level_domain is a good starting point

File: lighthouse-core/lib/url-shim.js [hunk 4]
```diff
@@ -17,6 +17,24 @@ const Util = require('../report/html/renderer/util.js');
 const URL = /** @type {!Window["URL"]} */ (typeof self !== 'undefined' && self.URL) ||
     require('url').URL;
 
+const tldPlusOne = {
```
Comment (paulirish) [inline lighthouse-core/lib/url-shim.js]: let's hardcode just the keys of this list (and add .co.uk)

add comment saying that https://publicsuffix.org/ is the real source of this data. and hypothetically we could pull in a dep like https://npms.io/search?q=public+suffix+list to do this forreal, but at significant byte cost.

and admit we're taking the lazy way out.

Comment (patrickhulce) [reply lighthouse-core/lib/url-shim.js]: shared a sheet with you on the most popular 2-level TLDs, but summary: if we did a regex with /\.(co|com|org|edu|gov)\.TLD$/ that should cover most cases

Comment (wardpeet) [reply lighthouse-core/lib/url-shim.js]: I added a new list to the regex coming from http archive. Not sure how I should comment it in the code. Also i'm not matching all 1000 entries

File: lighthouse-core/lib/url-shim.js [hunk 5]
```diff
@@ -17,6 +17,13 @@ const Util = require('../report/html/renderer/util.js');
 const URL = /** @type {!Window["URL"]} */ (typeof self !== 'undefined' && self.URL) ||
     require('url').URL;
 
+const tldPlusOneRegex = new RegExp(`\\.(com|co|gov|org|ac|or|edu|go|nic|in|gouv|gob|web|spb|net|` +
```
Comment (patrickhulce) [inline lighthouse-core/lib/url-shim.js]: I think we can trim down this list a bit based on the coverage percentages in that sheet, doing just the top 10 already gives us 94% coverage

```
com       2321232
co        1714813
gov       435538
edu       304490
ac        222746
org       202231
go        82772
gob       73721
or        72214
net       43587
----------------
sum       5473344
total     5841097
coverage  ~94%

File: lighthouse-core/lib/url-shim.js [hunk 6]
```diff
@@ -87,6 +94,55 @@ class URLShim extends URL {
...
+  /**
+   * Gets the tld of a domain
+   *
+   * @param {string} hostname
+   * @return {string} tld
+   */
+  static getTld(hostname) {
+    const tld = hostname.split('.').slice(-1)[0];
+
+    const tldRegex = new RegExp(`${tldPlusOneRegex.source}\\.${tld}$`);
```
Comment (patrickhulce) [inline lighthouse-core/lib/url-shim.js]: let's just do the `.split` and a `tlds.has` instead :)

**Comment by wardpeet** [comment]
Updated not sure this is what you had in mind

File: lighthouse-core/test/audits/uses-rel-preload-test.js [hunk 7]
```diff
@@ -50,14 +50,18 @@ describe('Performance: uses-rel-preload audit', () => {
 
   it('should suggest preload resource', () => {
     const rootNode = buildNode(1, 'http://example.com');
-    const mainDocumentNode = buildNode(2, 'http://www.example.com');
+    const mainDocumentNode = buildNode(2, 'http://www.example.com:3000');
```
Comment (patrickhulce) [inline lighthouse-core/test/audits/uses-rel-preload-test.js]: we're gonna need some graph creation helpers soon to make all this stuff :)

File: lighthouse-core/test/audits/uses-rel-preload-test.js [hunk 8]
```diff
@@ -69,41 +73,63 @@ describe('Performance: uses-rel-preload audit', () => {
...
+          nodeTimings.set(scriptOtherNodeLocal, {startTime: 1000, endTime: 2500});
+        }
+
+        return {timeInMs: 3500, nodeTimings};
       },
     };
 
     const mainResource = Object.assign({}, defaultMainResource, {
-      url: 'http://www.example.com',
+      url: 'http://www.example.com:3000',
```
Comment (patrickhulce) [inline lighthouse-core/test/audits/uses-rel-preload-test.js]: is this just to exercise extra hostname checks?

File: lighthouse-core/test/audits/uses-rel-preload-test.js [hunk 9]
```diff
@@ -131,7 +165,7 @@ describe('Performance: uses-rel-preload audit', () => {
     return UsesRelPreload.audit(mockArtifacts(networkRecords, chains, mainResource), {}).then(
       output => {
         assert.equal(output.rawValue, 1250);
-        assert.equal(output.details.items.length, 1);
+        assert.equal(output.details.items.length, 2);
```
Comment (patrickhulce) [inline lighthouse-core/test/audits/uses-rel-preload-test.js]: maybe overkill but assert the URL matches too?

File: lighthouse-core/lib/url-shim.js [hunk 10]
```diff
@@ -17,6 +17,10 @@ const Util = require('../report/html/renderer/util.js');
 const URL = /** @type {!Window["URL"]} */ (typeof self !== 'undefined' && self.URL) ||
     require('url').URL;
 
+const listOfTlds = [
```
Comment (patrickhulce) [inline lighthouse-core/lib/url-shim.js]: Add a comment that somewhat documents what happened here re: coverage and whatnot? maybe just link to the best comment thread on this PR :)

File: lighthouse-core/test/audits/uses-rel-preload-test.js (hunk 8 above)
Comment (wardpeet) [reply lighthouse-core/test/audits/uses-rel-preload-test.js]: yes it's to make sure that origins with a port number are handled correct by the rootDomainMatch

**Comment by patrickhulce** [comment]
so I just realized in #5471, we don't want to hide URLs based on if *they* were 3rd party, we want to hide them if *their initiator* was a 3rd party 🤔 nearly all of this should still apply, but I'll prob make an update soon :)

**Pull request merged**